py/src/call.py

```diff
@@ -0,0 +1,427 @@
+    return int(s, 10)
+
+
+def len_safe_hex(s):
+    """
+    Over at the RPC side which pathfinder supports, sometimes hex without
+    leading zeros is needed, so they could come over to python as well.
+    bytes.fromhex doesn't support odd length hex, it gives a non-hex character
+    error.
+    """
+    if s.startswith("0x"):
+        s = s[2:]
+    if len(s) % 2 == 1:
+        s = "0" + s
+    return bytes.fromhex(s)
+
+
+def string_or_int(s):
+    if type(s) == int:
+        return s
+
+    if type(s) == str:
+        if s.startswith("0x"):
+            return hash_or_int(s)
+        # not sure if this should be supported but strings get ran through the
+        # truncated keccak
+        return s
+
+    raise TypeError(f"expected string or int, not {type(s)}")
+
+
+def list_of_hash_or_int(s):
+    assert type(s) == list, f"Expected list, got {type(s)}"
+    return list(map(hash_or_int, s))
+
+
+def check_schema(connection):
+    global first
+    assert connection.in_transaction
+    cursor = connection.execute("select user_version from pragma_user_version")
+    assert cursor is not None, "there has to be an user_version defined in the database"
+
+    [version] = next(cursor)
+    return version == 2
+
+
+def resolve_block(connection, at_block):
+    from starkware.starknet.business_logic.state import BlockInfo
+
+    # FIXME: at least latest is wrong, and will be more wrong in the upcoming changes to schema
+
+    if at_block == "latest":
+        cursor = connection.execute(
+            "select starknet_block_number, starknet_block_timestamp, starknet_global_root from global_state order by starknet_block_number desc limit 1"
+        )
+    elif type(at_block) == int:
+        cursor = connection.execute(
+            "select starknet_block_number, starknet_block_timestamp, starknet_global_root from global_state where starknet_block_number = ?",
+            [at_block],
+        )
+    else:
+        assert type(at_block) == bytes, f"expected bytes, got {type(at_block)}"
+        if len(at_block) < 32:
+            # left pad it, as the fields in db are fixed length for this occasion
+            at_block = b"\x00" * (32 - len(at_block)) + at_block
+
+        cursor = connection.execute(
+            "select starknet_block_number, starknet_block_timestamp, starknet_global_root from global_state where starknet_block_hash = ?",
+            [at_block],
+        )
+
+    try:
+        [(block_number, block_time, global_root)] = cursor
+        return (
+            BlockInfo(block_number, block_time),
+            global_root,
+        )
+    except Exception:
+        raise NoSuchBlock(at_block)
+
+
+class NoSuchBlock(Exception):
+    def __init__(self, at_block):
+        super().__init__(f"Could not find the block by: {at_block}")
+
+
+class NoSuchContract(Exception):
+    def __init__(self):
+        super().__init__("Could not find the contract")
+
+
+class UnexpectedSchemaVersion(Exception):
```

Comment:
Same as above: expected vs found schema?

Reply:
Resolving cause YAGNI, feel free to unresolve and just ping me.